[Feat] 세계시간 페이지 mvp 구현 95%

seongto opened the pull request:
## � 요약(Summary)

세계시간 페이지의 기본적인 추가 및 시계 확인 기능 구현 완료.
mvvm 모델에서 rx를 열심히 써보려고 노력하였습니다.
다른 분들 작업한 코드들 보며 어느정도 일치화시키기위해 열심히 살펴보고 공부하였습니다만,
이해가 완전히 되지 않는 부분도 많아서 아는 부분까지 적당히 절충해가며 코드를 짜보았습니다.

## �️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [x] 버그 수정
- [x] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [x] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제


![Simulator Screen Recording - iPhone 16 Pro - 2025-01-13 at 18 07 20](https://github.com/user-attachments/assets/f4b82095-18d7-403c-88e8-010450a6be40)

위 이미지는 괴상한 코드로 작성했을 때의 결과물입니다. 좀 기다리시면 이상한 장면을 볼 수 있어요.

## � 공유사항 to 리뷰어

1. Inject 관련 현재 에러가 있습니다. 본래 개발자 취지와 다르게 일어나는 문제라 해결책을 고심하는 중입니다.
2. 마지막 괴상한 코드를 걷어내는 중 실시간 시간 업데이트에 문제가 살짝 발생하였습니다. 해당 이슈 해결을 최우선으로 하고 있습니다.
3. 삭제 기능을 어떤 모습으로 구현할지 고민 중입니다. 스와이프 제스쳐로 시도했으나 실패하였습니다. 삭제 자체는 어렵지 않은데 그걸 어떻게 구현할지가 관건이네요.

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

seongto committed:
[Feat] 구분선 표시용 UIView 추가

seongto committed:
[Merge] 브랜치 최신화 병합

seongto committed:
[Feat] 세계 시간 UI 작업 완료

- 모달뷰 처리 방식 변경
- 세계시간 페이지의 기본적인 UI 및 이벤트 처리틀 완료

seongto committed:
[Merge] 브랜치 최신화 병합

branch 'develop' of https://github.com/1aborInModeration/OneMoreMinute into feature/world-ui

seongto committed:
[Feat] 세계시간 관련 코어데이터 설정 업데이트 외

- 코어 데이터 관련 내부 메소드와 속성이름 수정
- 공통 디자인 관련 시계 폰트 및 상수 속성 수정

seongto committed:
[Feat] 세계시간 UI mvp 완료

- 원하는 도시 검색 및 추가 기능 구현
- 도시 데이터는 코어데이터에 저장하여 언제든지 볼수 있음.
- 현재 시간 업데이트 관련 괴상한 로직이 적용되어 있습니다.
- 삭제 제스쳐는 추가하였으나, 현재 시간 업데이트 관련 로직 이슈로 문제.

seongto committed:
[Merge] 깃 브랜치 최신화 병합

branch 'develop' of https://github.com/1aborInModeration/OneMoreMinute into feature/world-ui

seongto committed:
[Feat] 세계시계 1초 초기화 수정 및 제스쳐 삭제

- 1초마다 초기화하는 로직을 변경하였으나, 타이머 정상작동 안하는 중.
- 스와이프 제스처를 이용한 삭제버튼 활성화 노력이 실패하여 삭제.
- 현재 삭제 기능이 없습니다.

seongto commented:
크아... 삽질과 가지치기 많이 하다가 너무 한방에 올려서 참 죄송합니다.... 제가 하자고 권장했던 것과 너무 다르게 일했네요...
근데... 진짜... 아니... 이 페이지 진짜... 너무 힘들었어요... 아니.. 생각보다 왜 할게 많죠?

seongto committed:
[Merge] 브랜치 최신화 병합

branch 'develop' of https://github.com/1aborInModeration/OneMoreMinute into feature/world-ui

seongto commented:
아이콘 및 런치화면 관련 브랜치 최신화 포함하였습니다.

ericKwon95 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/ViewModel/SearchCityViewModel.swift:
```diff
@@ -16,39 +16,81 @@ class SearchCityViewModel {
+
+
+extension SearchCityViewModel {
+    private func getTimeZone(from coordinate: CLLocationCoordinate2D) -> TimeZone? {
```
타임존 관련 로직 잘 보고 갑니다!

ericKwon95 reviewed (approved):
고생 많으셨습니다...!

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/Core/Theme/Fonts.swift:
모노스페이스 폰트는 어떻게 쓰는건가요?
저는 써본 적이 없어서 사용 의도와 사용법이 궁금합니다!!

NeoSelf1 commented on OneMoreMinute/OneMoreMinute/Source/View/UIComponent/UILabel/ClockLabel.swift:
```diff
@@ -0,0 +1,82 @@
+}
+
+class ClockLabel: UILabel {
+    // MARK: - Life Cycles
```
@seongto님, 줄넘김 기입 부탁드리겠습니다!!

seongto replied:
아 Mark 넣을 때도 적용해야 하는거군요-! 넣어두도록 하겠습니다!

NeoSelf1 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/Model/WorldTime.swift:
```diff
@@ -0,0 +1,15 @@
+//  Created by MaxBook on 1/13/25.
+//
+
+import Foundation
```
@seongto 님, 혹시 Foundation을 import한 이유가 무엇일까요??

seongto replied:
습관성.....?

DoyleHWorks commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/View/WorldTimeCell.swift:
```diff
@@ -0,0 +1,171 @@
+
+// MARK: - Gesture Recognizers
+
+extension WorldTimeCell {
+    /// 컨텐츠 에어리어가 터치되었을 때마다 삭제 버튼 토글
+    func contentsAreaTapped() {
+        isDeleteButtonVisible = !isDeleteButtonVisible
+        deleteButton.isHidden = !isDeleteButtonVisible
+        
+        self.contentWrapperView.snp.makeConstraints { make in
+            if isDeleteButtonVisible {
+                make.trailing.equalToSuperview().inset(Layouts.buttonHeight + Layouts.itemSpacing1)
+            } else {
+                make.trailing.equalToSuperview()
+            }
+        }
+    }
+
+    
+    // MARK: - Button Action
+    func deleteButtonTapped() {
+        print("삭제 버튼 클릭!")
+    }
+}
```
형석님이 스크럼에서 말씀하셨던 것처럼, 별도 메서드를 찾아서 구현하면 더 좋을 것 같습니당
조금 찾아보니까 이런 리소스도 있네요:
https://nemecek.be/blog/164/how-to-add-swipe-actions-to-collection-view

NeoSelf1 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/View/CityListCell.swift:
```diff
@@ -0,0 +1,75 @@
+    
+}
+
+
```
@seongto 님, 줄넘김 확인 부탁드리겠습니다!!

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/RepositoryManager/WorldTimeDataManager/WorldTimeDataManager.swift:
```diff
@@ -34,6 +34,9 @@ final class WorldTimeDataManager: CoreDataManaged {
     /// - Returns: Alarm 엔티티의 데이터 목록 배열
     func fetch() -> [Entity] {
         let request: NSFetchRequest<Entity> = Entity.fetchRequest()
+        let sort = NSSortDescriptor(key: "createdAt", ascending: false)
+        request.sortDescriptors = [sort]
+        
```
정렬을 위해 이런 코드를 추가하셨군요! 좋습니다ㅎㅎ

DoyleHWorks reviewed (approved):
고생하셨습니다!

NeoSelf1 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/View/SearchCityView.swift:
```diff
@@ -0,0 +1,122 @@
+
+// MARK: - Update UI
+
+extension SearchCityView {
```
@seongto 님, 혹시 extension은 추후 구현될 로직을 위해 미리 만들어두신건가요??

seongto replied:
아 제가 UIView 랑 ViewController 는 양식화 시켜놓고 쓰는 부분이 있어서, 아직은 비어있는 저런것도 있습니다.
추후 문서화주석 작업과 함께 컨벤션 부분 적용 하도록 하겠습니다!

NeoSelf1 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/ViewController/WorldTimeViewController.swift:
```diff
@@ -0,0 +1,84 @@
+            }
+            .disposed(by: disposeBag)
+        
+//        worldTimeView.worldTimeCollectionView.rx
```
@seongto 님, 만일 사용하지 않는 코드면 커밋내역에 포함시키지 않는 것도 좋아보여요!

NeoSelf1 reviewed (changes requested):
@seongto 님, 고생하셨습니다!! 시간 괜찮으시다면 MARK 주석 전후 개행 컨벤션 및 비어있는 extension 관련하여 검토 부탁드리겠습니다..!

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/View/UIComponent/UIButton/CellGestureButton.swift:
```diff
@@ -0,0 +1,48 @@
+
+extension CellGestureButton {
+    func setupUIProperties(type: gestureButtonType) {
+        self.translatesAutoresizingMaskIntoConstraints = false
```
`SnapKit`을 사용하셨는데, 이 코드를 사용하는 이유가 있을까요??

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/View/UIComponent/UILabel/ClockLabel.swift:
```diff
@@ -0,0 +1,82 @@
+            self.textAlignment = .center
+        }
+    }
+}
```
enum으로 레이블 타입을 관리해서 케이스별로 다른 값을 부여해주는 방법 좋네요~
역시 공용 컴포넌츠 관리자...

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/View/UIComponent/UISearchBar/customSearchBar.swift:
서치바가 있으니 키보드가 있겠군요!!
성수님도 키보드를 내리는 방법, 키보드가 올라왔을 때 뷰를 가린다면 그 해결방법 등을 찾아보시면 좋을 것 같아요!!
힌트 코드는 `endEditing`입니다ㅎ ㅎ

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/View/SearchCityView.swift:
```diff
@@ -0,0 +1,122 @@
+
+// MARK: - Update UI
+
+extension SearchCityView {
+
+}
```
여긴 혹시 왜 비어있을까요?

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/View/SearchCityView.swift:
```diff
@@ -0,0 +1,122 @@
+
+// MARK: - Action
+
+extension SearchCityView {
+    func didTapButton() {
+        
+    }
+}
+
+
+
+// MARK: - Modal Porotocol Constraints
+
+extension SearchCityView {
+    func onModalWillAppear() {
+        // 모달이 나타날 때 동작 추가
+    }
+    
+    func onModalWillDisappear() {
+        // 모달이 사라질 때 동작 추가
+    }
+    
+    func closeModal() {
+        delegate?.closeModal()
+    }
+}
```
여기는 아직 미구현 기능인가요

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/ViewController/SearchCityModalViewController.swift:
```diff
@@ -0,0 +1,129 @@
+
+// MARK: - Action Management & Mapping
+
+extension SearchCityModalViewController {
+    func closeModal() {
+        self.dismiss(animated: true, completion: nil)
+    }
```
모달을 델리게이트 패턴으로 닫아주셨군요!!
Rx 바인딩으로도 구현할 수 있을까요?

Crois0509 commented on OneMoreMinute/OneMoreMinute/Source/View/WorldTime/ViewController/WorldTimeViewController.swift:
```diff
@@ -0,0 +1,84 @@
+    func setupBindings() {
+        worldTimeView.plusButton.rx
+            .tap
+            .subscribe(onNext: { [weak self] in
+                if let topVC = AppHelpers.getTopViewController() {
+                    let searchModalVC = Inject.ViewControllerHost(SearchCityModalViewController())
+                    searchModalVC.modalPresentationStyle = .overFullScreen
+                    searchModalVC.modalTransitionStyle = .crossDissolve
+                    topVC.present(searchModalVC, animated: true) {
+                        searchModalVC.dismissCompletion = { [weak self] in
+                            print("modal dismissed")
+                            self?.testPrint()
+                        }
+                    }
+                }
+            })
```
이 부분이 2중 클로저를 사용하며 네스팅이 깊어진 것 같아요.
클로저 사용을 줄여서 네스팅을 줄일 수 있는 방면으로 가보면 어떨까요?
그리고 `present`로 띄우는 뷰 컨트롤러는 항상 최상위 뷰로 표시되기 때문에, 굳이 topVC를 찾아서 넣지 않아도 될 것 같습니다!!
만약 백그라운드 뷰가 별도로 필요하다면 그 경우에만 넣어도 좋을 것 같아요!

`dismiss` 액션도 바인딩 등을 통해 별도로 구현할 수 있다고 생각합니다!

Crois0509 reviewed (approved):
구현 너무 고생하셨습니다!!
제 사적 의견이 잔뜩 담긴 코멘트를 몇 남겼으니 확인 부탁드려요!

seongto committed:
[Feat] 세계시간 컬렉션 뷰에 스와이프 액션 추가

- 세계시간 컬렉션 뷰에 스와이프 액션 추가방법을 구현하였습니다.
- 뷰컨트롤러에서 원하는 형태로 액션을 매핑하여 삽입 가능.
- indexPath를 사용할 수 있습니다.
- 자유롭게 원하는 개수, 원하는 방향(trailing, leading) 설정 가능.

seongto committed:
[Feat] 잘못된 주석 수정 및 빈줄 처리 추가

seongto committed:
[Feat] Inject 삭제

- instruments 사용을 방해하는 못된 Inject를 제거합니다.

seongto committed:
[Fix] Inject 라이브러리 삭제

seongto committed:
[Feat] 세계 시간 카드 삭제 기능 구현

seongto committed:
[Merge] 깃 브랜치 최신화 커밋

branch 'develop' of https://github.com/1aborInModeration/OneMoreMinute into feature/world-ui

seongto committed:
[Fix] 라이브러리 병합 충돌 찌꺼기 정리.

seongto commented:
### **추가 푸시 내역**

- 세계시간 컬렉션 뷰에 스와이프 액션 추가방법을 구현하였습니다.
- 뷰컨트롤러에서 원하는 형태로 액션을 매핑하여 삽입 가능.
- indexPath를 사용할 수 있습니다.
- 자유롭게 원하는 개수, 원하는 방향(trailing, leading) 설정 가능.
- 세계 시간 삭제 기능을 추가하였습니다.
- Inject 라이브러리 관련 코드를 모두 삭제하였습니다.

ericKwon95 commented on OneMoreMinute/OneMoreMinute/Source/View/UIComponent/UIButton/CustomFilledButton.swift:
```diff
@@ -19,9 +19,9 @@ class CustomFilledButton: UIButton  {
     // MARK: - Life Cycles
     
-    /// 헤드라인 라벨
+    /// 커스텀 filled 버튼
```
커밋 너무 좋아졌습니다!
요 커밋은 [Docs] 또는 [Chore]이 어울릴 것 같아요~!

ericKwon95 reviewed (approved):
커밋 아주 좋아졌어요!
커밋 태그 중 Feat이 아닌 Docs, Chore, Build 등 다양한 태그도 함께 활용해 보시면 매우 좋을 것 같습니다~!

NeoSelf1 reviewed (commented):
@seongto 님, 고생 많으셨습니다... 새로운 Swipe Action 구현 방법 많이 배웠습니다!!
